### ISSUE TITLE
How is the "teacher" mechanism realised?

### ISSUE THREAD

**Issue by Zephyr69** [orgilj/semi-yolov5#1]
Hi there! This is interesting but I don't get how the teacher model (or its equivalent mechanism) in the original STAC paper is implemented here?

**Comment by orgilj** [comment]
Thank you for your interest. There is no teacher model and paper. I followed the original semi-supervised learning method. First, we need to feed the model using unlabeled batch images and get out1, then feed labeled batch images and calculate the loss. After that, backward loss and feed previously used unlabeled batch get out2. Then calculate loss [unlabeled_loss = loss_fn(out1, ou2)] and backward it. After some iteration model will learn something from unlabeled data. You can do this with any model.

unlabeled_out1 = model(unlabeled_batch)
out = model(labeled_batch)
loss = loss_fn(out, target)
loss.bachward()
unlabeled_out2 = model(unlabeled_batch)
un_loss = loss_fn(unlabeled_out1, unlabeled_out2)
un_loss.backward()

**Comment by Zephyr69** [comment]
But in the code wouldn't these two outs produce exactly the same result, since the optimizer hasn't done anything?
[https://github.com/orgilj/semi-yolov5/blob/master/train.py#L338-L352](url)

**Comment by orgilj** [comment]
Thank you for your interest. Yes in this code need some changes. If you want I’ll send you result video.

**Comment by Zephyr69** [comment]
> Thank you for your interest. Yes in this code need some changes. If you want I’ll send you result video.
> […](#)
> Sent from my iPhone
> On 7 Nov 2022, at 21:26, Zephyr69 ***@***.***> wrote: ﻿ But in the code wouldn't these two outs produce exactly the same result, since the optimizer hasn't done anything? https://github.com/orgilj/semi-yolov5/blob/master/train.py#L338-L352 — Reply to this email directly, view it on GitHub, or unsubscribe. You are receiving this because you commented.

Sure! Thank you in advance!
